### PR TITLE
feat: Update OpenAI spec to include support for tools

### LINKED ISSUE #117: Feature: Add support to include tools for OpenAI Chat Template to enable function calling.
### Description

Introduce support for integrating tools in chat messages, similar to the [OpenAI GPT-4 function calling](https://platform.openai.com/docs/guides/function-calling). This feature would enable users to include tools in their chat interactions, allowing the model to return specific function based on user queries.

Example usage:

```json
{
  "model": "lit",
  "messages": [
    {
      "role": "user",
      "content": "What's the weather like in Boston today?"
    }
  ],
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "get_current_weather",
        "description": "Get the current weather in a given location",
        "parameters": {
          "type": "object",
          "properties": {
            "location": {
              "type": "string",
              "description": "The city and state, e.g. San Francisco, CA"
            },
            "unit": {
              "type": "string",
              "enum": ["celsius", "fahrenheit"]
            }
          },
          "required": ["location"]
        }
      }
    }
  ],
  "tool_choice": "auto"
}
```

This proposal also includes adding options for `Tool Message` in messages and `tool_calls` in the assistant message.

### Additional Context

This feature would streamline the implementation of chat generation APIs with function calling support, aligning with OpenAI's specifications and benefiting users seeking more interactive and functional chat applications.

### PR DESCRIPTION
**Pull request by bhimrazy**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #117.
- Add Support for Tools in the OpenAI Spec Template
- Also updates default chat message class to include `name`, `tool_calls` & `tool_calls_id`

### CODE REVIEW

**Comment by bhimrazy** [comment]
Hi @aniketmaurya,

I wanted to check about the ChatMessage class.

Is it okay to have the optional fields in the same class, or is it necessary to split it into separate classes like UserMessage, AssistantMessage, ToolMessage, etc.?

While mistral_common follows the split format, the Function Calling documentation at https://docs.mistral.ai/capabilities/function_calling/ has all of the fields in the ChatMessage class.

Here is the relevant GitHub link: https://github.com/mistralai/client-python/blob/main/src/mistralai/models/chat_completion.py.

Also, I was looking for how `tool_choice` is used & how `ids` are assigned to function suggested by the llms.

Thank you!
cc: @lantiga

**Comment by aniketmaurya** [comment]
Thanks for the PR @bhimrazy!! I am taking a look 👀

File: src/litserve/specs/openai.py [hunk 1]
```diff
@@ -52,9 +55,40 @@ class ImageContent(BaseModel):
...
+    function: Function
+
+
+class FunctionCall(BaseModel):
+    name: str
+    arguments: str
+
+
+class ToolCall(BaseModel):
+    id: str = "null"
```
Comment (aniketmaurya) [inline src/litserve/specs/openai.py:81]:
```suggestion
    id: Optional[str] = None
```

File: src/litserve/specs/openai.py [hunk 2]
```diff
@@ -74,6 +108,8 @@ class ChatCompletionRequest(BaseModel):
     presence_penalty: Optional[float] = 0.0
     frequency_penalty: Optional[float] = 0.0
     user: Optional[str] = None
+    tools: Optional[List[Tool]] = None
```
Comment (aniketmaurya) [inline src/litserve/specs/openai.py:111]: this should be `tool_calls`. I sent the sample request to openai with streaming and get the following JSON response. 


```
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":"assistant","tool_calls":null},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":"call_kz3cshGHfs03xZAbjpRwyDTE","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"n\": \"S","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"an F","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"ranci","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"sco, C","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"A\", ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"\"unit","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"\": \"ce","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"lsiu","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"s\"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":"call_jmYzksaRe2oms7Jp3Vuj3EyK","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"n\": \"T","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"okyo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"\", \"u","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"nit\": ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"\"cel","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"sius\"","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":"call_9VH5q6ssHQuiZ41W1vmZqU2b","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"n\": \"P","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"aris","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"\", \"u","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"nit\": ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"\"cel","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"sius\"","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":null},"finish_reason":"tool_calls","index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
```

**Review by aniketmaurya — CHANGES_REQUESTED** [review]
Looks great! Added some comments so that we align with OpenAI specs.

**Comment by aniketmaurya** [comment]
Also, pasting the non-streaming response from openai for reference @bhimrazy  - 

```
{
    "id": "chatcmpl-9UWuKx5zkBrbGpaN8iNnCYiBWtoSn",
    "choices": [
        {
            "finish_reason": "tool_calls",
            "index": 0,
            "logprobs": null,
            "message": {
                "content": null,
                "role": "assistant",
                "function_call": null,
                "tool_calls": [
                    {
                        "id": "call_vbuFjXcjrereRNEDWI43s",
                        "function": {
                            "arguments": "{\"location\": \"San Francisco, CA\", \"unit\": \"celsius\"}",
                            "name": "get_current_weather"
                        },
                        "type": "function"
                    },
                    {
                        "id": "call_A738j1sExArrewZtiOvH",
                        "function": {
                            "arguments": "{\"location\": \"Tokyo, Japan\", \"unit\": \"celsius\"}",
                            "name": "get_current_weather"
                        },
                        "type": "function"
                    },
                    {
                        "id": "call_6vnreerMz2yCjNLs0Qo",
                        "function": {
                            "arguments": "{\"location\": \"Paris, France\", \"unit\": \"celsius\"}",
                            "name": "get_current_weather"
                        },
                        "type": "function"
                    }
                ]
            }
        }
    ],
    "created": 1717063364,
    "model": "gpt-4o-2024-05-13",
    "object": "chat.completion",
    "system_fingerprint": "fp_43dfreef1",
    "usage": {
        "completion_tokens": 83,
        "prompt_tokens": 85,
        "total_tokens": 168
    }
}
```

**Comment by lantiga** [comment]
> Hi @aniketmaurya,
> 
> I wanted to check about the ChatMessage class.
> 
> Is it okay to have the optional fields in the same class, or is it necessary to split it into separate classes like UserMessage, AssistantMessage, ToolMessage, etc.?
> 
> While mistral_common follows the split format, the Function Calling documentation at https://docs.mistral.ai/capabilities/function_calling/ has all of the fields in the ChatMessage class.
> 
> Here is the relevant GitHub link: https://github.com/mistralai/client-python/blob/main/src/mistralai/models/chat_completion.py.
> 
> Also, I was looking for how `tool_choice` is used & how `ids` are assigned to function suggested by the llms.
> 
> Thank you! cc: @lantiga

Great contribution @bhimrazy! I personally don't have a strong preference with respect to separating classes, as long as

- the produced json is equivalent
- it's easy for users to figure out how to pack/unpack <- this is actually the main driver IMO

I find having a single ChatMessage is easier to grasp, but let's merge this and see if users are confused by it.

File: src/litserve/specs/openai.py (hunk 2 above)
Comment (bhimrazy) [reply src/litserve/specs/openai.py:111]: Thank you @aniketmaurya for the sample response. It was really helpful.

BTW, In the above suggestions, `tools` is for ChatCompletionRequest, to receive available tools input.
I  have already added `tool_calls` in the chat message.
```python
class ChatMessage(BaseModel):
    role: str
    content: Union[str, List[Union[TextContent, ImageContent]]]
    name: Optional[str] = None
    tool_calls: Optional[List[ToolCall]] = None
    tool_call_id: Optional[str] = None
```

**Comment by bhimrazy** [comment]
Hi @aniketmaurya and @lantiga, thank you for all the feedback. 
The PR is ready for review. 
Thanks.

Additional Notes:
- One of the tests is failing, which is caused by the README example. I'm not sure how to fix it.
- Also, during manual local testing, the processes are not killed if the test fails. Is it an issue?

**Review by aniketmaurya — COMMENTED** [review]
thank you so much @bhimrazy! Hope you had fun while implementing the PR as well. Looks good to me 😄

**Comment by bhimrazy** [comment]
> thank you so much @bhimrazy! Hope you had fun while implementing the PR as well. Looks good to me 🚀 .
> 
> Regarding the processes not getting killed on test fails, I have created an issue [here](https://github.com/Lightning-AI/LitServe/issues/121). Let me know if you wanna take a stab at that 😄

Sure, Thank you.

**Review by aniketmaurya — APPROVED** [review]

**Comment by bhimrazy** [comment]
> Great job once again @bhimrazy, let's ship this! 🚀

Thank you, @lantiga and @aniketmaurya. 
I can't believe this is my second PR getting merged. 
I'm excited to keep contributing more!

**Pull request merged**